### PR TITLE
Representers

### PR DESCRIPTION
**Pull request by Bugagazavr**
This PR contains:
 - allow to attach custom model parser via `GrapeSwagger.register_model_parser`
 - trailblazer representers support
 - model parsers moved to separated classes
 -  grape-entity not a runtime dependency

TODO:
 - changelog
 - update README

### CODE REVIEW

**Comment by Bugagazavr** [comment]
Also related:

- https://github.com/ruby-grape/grape-swagger/issues/121

This allows to create gem with custom model parsers, like `representable`, `roar`, `grape-entity`

@dblock

File: lib/grape-swagger/models/representable.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        representer.map.each_with_object({}) do |value, property|
+          property[value.name] = parse_representer_property(value)
+        end
+      end
+    end
+  end
+end
+
+if defined?(::Representable)
+  GrapeSwagger.register_model_parser(GrapeSwagger::Models::Representable, ::Representable::Decorator)
```
Comment (dblock) [inline lib/grape-swagger/models/representable.rb]: I think this could be more elegant with something like `GrapeSwagger.model_parsers.register ...`.

**Comment by dblock** [comment]
Overall, big 👍 on this. Looking forward to seeing this finished.

We could take grape-entity and representable parts out of grape-swagger altogether into separate gems instead of keeping support here or not.

**Comment by Bugagazavr** [comment]
@dblock im add model parsers collection, but many specs uses grape entities in test.

I create two gems:

- https://github.com/Bugagazavr/grape-swagger-representable
- https://github.com/Bugagazavr/grape-swagger-entity

Should i publish this gems before this merged?

**Comment by dblock** [comment]
I think you should and make it very clear in UPGRADING in grape-swagger what users are supposed to do if they use one or the other. I would move specific tests into those projects and leave some basic tests that work for either library in this PR. You could also have a `.travis.yml` that sets something like `ENTITY_LIBRARY` and runs "integration" tests with it, a little bit how https://github.com/dblock/slack-ruby-client for example switches between faye-websocket and celluloid to test.

**Comment by dblock** [comment]
We don't have to, but if you would like to move grape-swagger-* into the ruby-grape organization we can do that too. I can make you co-maintainer of all three, grape-swagger and those two.

**Comment by Bugagazavr** [comment]
@dblock ok, im agreed.

I squash my commits, add some information to readme and update changelog.

For specs uses `grape-swagger-entity`

`.travis.yml` also updated in https://github.com/Bugagazavr/grape-swagger-representable and https://github.com/Bugagazavr/grape-swagger-entity

So, now if this this is not problem, we can move this repos to ruby-grape namespace, and delete git ref from Gemfiles to my `grape-swagger` and add dependencies to gempecs

**Comment by Bugagazavr** [comment]
ah, ok i forgot add `ENTITY_LIBRARY`, take a moment i add this

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 * [#405](https://github.com/ruby-grape/grape-swagger/pull/405), [#403](https://github.com/ruby-grape/grape-swagger/issues/403): Added version support matrix - [@LeFnord](https://github.com/LeFnord).
 * [#408](https://github.com/ruby-grape/grape-swagger/pull/408): Added support for `HEAD` endpoints - [@Bugagazavr](https://github.com/Bugagazavr).
 * [#408](https://github.com/ruby-grape/grape-swagger/pull/411): Added support for `OPTIONS` endpoints - [@Bugagazavr](https://github.com/Bugagazavr).
+* [#413](https://github.com/ruby-grape/grape-swagger/pull/413): Move all model parsing logic to separate gems `grape-swagger-entity` and added representable parser `grape-swagger` - [@Bugagazavr](https://github.com/Bugagazavr)
```
Comment (dblock) [inline CHANGELOG.md]: Missing a period at the end :)

File: .travis.yml [hunk 3]
```diff
@@ -5,6 +5,8 @@ sudo: false
 matrix:
   include:
     - rvm: 2.3.1
+      env: ENTITY_LIBRARY=representable
```
Comment (dblock) [inline .travis.yml]: And you want another one for grape-entity, right?

File: Gemfile [hunk 4]
```diff
@@ -8,3 +8,6 @@ when 'HEAD'
 else
   gem 'grape', version
 end
+
+gem 'grape-swagger-entity', github: 'bugagazavr/grape-swagger-entity'
```
Comment (dblock) [inline Gemfile]: I think the ENV variable should be `MODEL_PARSER` to be consistent, should be set to either `grape-swagger-entity` or `grape-swagger-representable` and be `gem ENV['MODEL_PARSER'] if ENV.key?('MODEL_PARSER')`. The specs should pass without a model parser library, I don't think we should have a default and `.travis.yml` would have two entries to make things very explicit.

File: README.md [hunk 5]
```diff
@@ -27,12 +27,24 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
 
 These screenshot is based on the [Hussars](https://github.com/LeFnord/hussars) sample app.
 
+## Model Parsers
+
+Since 0.21.0, `Grape::Entity` is not a part of grape-swagger, you need to add `grape-swagger-entity` manually to your Gemfile:
```
Comment (dblock) [inline README.md]: Separate `grape-swagger-representable` and add something like "Grape-swagger also supports Representable. To ... add ...".

File: grape-swagger.gemspec [hunk 6]
```diff
@@ -12,9 +12,10 @@ Gem::Specification.new do |s|
   s.license     = 'MIT'
 
   s.add_runtime_dependency 'grape', '>= 0.12.0'
-  s.add_runtime_dependency 'grape-entity'
   s.add_runtime_dependency 'awesome_print'
 
+  s.add_development_dependency 'grape-entity'
+  s.add_development_dependency 'representable'
```
Comment (dblock) [inline grape-swagger.gemspec]: These should be removed, now that they are enabled by ENV.

File: spec/lib/model_parsers_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
+require 'spec_helper'
+
+describe GrapeSwagger::ModelParsers do
+  let(:model_parsers) { described_class.new }
+
+  before do
+    class SomeModelParser; end
+    class SomeModelParser2; end
+    class SomeModelParser3; end
```
Comment (dblock) [inline spec/lib/model_parsers_spec.rb]: These are global, I think you should use `let(:parser1) { Class.new }`.

File: spec/support/representable_api_swagger_v2_result.rb [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            'others' => { 'description' => '', 'type' => 'text' }
+          },
+          'description' => 'This gets Things.'
+        } } }
+  end
+
+  let(:http_verbs) { %w(get post put delete) }
+end
+
+def mounted_paths
```
Comment (dblock) [inline spec/support/representable_api_swagger_v2_result.rb]: Beware of global functions! Should be refactored into a `let`.

**Comment by dblock** [comment]
This is very close, see my smallish comments. I don't think we should treat grape-entity as "default" in specs basically. In the end travis would run general specs for many versions of Ruby and only once against each entity library on top of the latest ruby version or something like that.

**Comment by dblock** [comment]
I'm going to cut a release of grape-swagger with whatever changes we have on HEAD now FYI, that should give us more time to announce/release this big change.

File: README.md [hunk 9]
```diff
@@ -46,7 +58,7 @@ grape-swagger | swagger spec | grape                   | grape-entity
 0.10.5        |     1.2      | >= 0.10.0 ... <= 0.14.0 |  < 0.5.0
 0.11.0        |     1.2      |               >= 0.16.2 |  < 0.5.0
 0.20.1        |     2.0      | >= 0.12.0 ... <= 0.14.0 | <= 0.5.1
-0.20.3 (next) |     2.0      | >= 0.12.0 ... <= 0.16.2 | <= 0.5.1
+0.21.0 (next) |     2.0      | >= 0.12.0 ... <= 0.16.2 | <= 0.5.1
```
Comment (dblock) [inline README.md]: Add a column here for representable.

File: .travis.yml (hunk 3 above)
Comment (Bugagazavr) [inline .travis.yml]: grape-entity is default

Comment (dblock) [inline .travis.yml]: I am saying we shouldn't have a default.

**Comment by Bugagazavr** [comment]
Ok, i create a simple mock shared example to launch tests without any entity library, and add 2 shared examples to tests grape-swagger with representable and grape-entity.

File: README.md [hunk 10]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+Also added support for [representable](https://github.com/apotonick/representable) via `grape-swagger-representable`.
+
+```ruby
+# For Grape::Entity ( https://github.com/ruby-grape/grape-entity )
+gem 'grape-swagger-entity'
+# For representable ( https://github.com/apotonick/representable )
+gem 'grape-swagger-representable'
+```
+
+##### Custom model parser interface
```
Comment (dblock) [inline README.md]: Make capitalization consistent, just "Custom Model Parsers"

File: README.md [hunk 11]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+
+You can create your own model parser, for example for [roar](https://github.com/apotonick/roar)
```
Comment (dblock) [inline README.md]: Add a period.

File: README.md [hunk 12]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+
+      def call
+        # Parse your model and return hash with model schema for swagger
+      end
+    end
+  end
+end
+```
+
+Then you should register your custom parser:
```
Comment (dblock) [inline README.md]: Change ":" to a period, just for consistency.

File: README.md [hunk 13]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+
+```rb
+GrapeSwagger.model_parsers.register(GrapeSwagger::Roar::Parser, Roar::Decorator)
+```
+
+To controll model parsers sequence, you can insert your Parser before or after another parser
```
Comment (dblock) [inline README.md]: Spelling: control with one l.

Make parser lowercase, add a period in the end.

File: README.md [hunk 14]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+
+before:
+
+```rb
+if defined?(GrapeSwagger::Representable::Parser)
```
Comment (dblock) [inline README.md]: I am confused by this `if`. I think for simplicity this should just be the before or after example.

File: README.md [hunk 15]
```diff
@@ -27,12 +27,79 @@ The grape-swagger gem provides an autogenerated documentation for your [Grape](h
...
+  GrapeSwagger.model_parsers.insert_before(GrapeSwagger::Representable::Parser, GrapeSwagger::Roar::Parser, Roar::Decorator)
+else
+  GrapeSwagger.model_parsers.register(GrapeSwagger::Roar::Parser, Roar::Decorator)
+end
+```
+
+after:
```
Comment (dblock) [inline README.md]: Maybe "#### insert_after"

File: README.md [hunk 16]
```diff
@@ -41,12 +108,13 @@ These screenshot is based on the [Hussars](https://github.com/LeFnord/hussars) s
...
 
-grape-swagger | swagger spec | grape                   | grape-entity
---------------|--------------|-------------------------|-------------
-0.10.5        |     1.2      | >= 0.10.0 ... <= 0.14.0 |  < 0.5.0
-0.11.0        |     1.2      |               >= 0.16.2 |  < 0.5.0
-0.20.1        |     2.0      | >= 0.12.0 ... <= 0.14.0 | <= 0.5.1
-0.20.3        |     2.0      | >= 0.12.0 ... ~> 0.16.2 | ~> 0.5.1
+grape-swagger | swagger spec | grape                   | grape-entity | representable |
+--------------|--------------|-------------------------|--------------|---------------|
+0.10.5        |     1.2      | >= 0.10.0 ... <= 0.14.0 |  < 0.5.0     | Unsupported   |
```
Comment (dblock) [inline README.md]: lowercase unsupported or make it n/a

**Comment by dblock** [comment]
I made a bunch of nitpick comments in README.

Could you please update https://github.com/ruby-grape/grape-swagger/blob/master/UPGRADING.md?

I'll merge then. Thanks for the excellent work.

**Comment by Bugagazavr** [comment]
@dblock done

**Comment by dblock** [comment]
Fix the build pls.

File: Gemfile [hunk 17]
```diff
@@ -8,3 +8,5 @@ when 'HEAD'
 else
   gem 'grape', version
 end
+
+gem ENV['MODEL_PARSER'], github: "bugagazavr/#{ENV['MODEL_PARSER']}" if ENV.key?('MODEL_PARSER')
```
Comment (dblock) [inline Gemfile]: Lets move the "bugagazavr/" part into the ENV variable?

File: UPGRADING.md [hunk 18]
```diff
@@ -1,6 +1,15 @@
 Upgrading Grape-swagger
 =======================
 
+### Upgrading to >= 0.21.0
+
+With grape >= 0.21.0, `grape-entity` support moved to separate gem `grape-swagger-entity`, if you use grape entity, update your Gemfile:
```
Comment (dblock) [inline UPGRADING.md]: "grape entity" -> "grape-entity"

**Comment by dblock** [comment]
Will hit merge on 💚.

**Comment by dblock** [comment]
I see https://github.com/ruby-grape/grape-swagger/pull/415, this is cool, I'll just merge and deal with it.

**Comment by dblock** [comment]
Feel free to address any of my minor comments in a future PR. You should e-mail the grape mailing list about this change asking people to give it a shot.

**Comment by dblock** [comment]
Also release the representer libraries and change the MODEL_PARSER env to the gem name instead of a github reference.

**Comment by Bugagazavr** [comment]
Thanks 👍  

https://rubygems.org/gems/grape-swagger-entity
https://rubygems.org/gems/grape-swagger-representable

What about move this gems to `ruby-grape` org?

**Comment by dblock** [comment]
@Bugagazavr lets do it offline

**Comment by dblock** [comment]
@Bugagazavr email me dblock at dblock dot org

**Pull request merged**